src/aggregator/integration/placement_change_test.go

```diff
@@ -0,0 +1,303 @@
+			SetInstrumentOptions(instrumentOpts).
+			SetElectionCluster(electionCluster).
+			SetHTTPAddr(mss.httpAddr).
+			SetInstanceID(initialInstanceConfig[i].instanceID).
+			SetKVStore(kvStore).
+			SetRawTCPAddr(mss.rawTCPAddr).
+			SetShardSetID(initialInstanceConfig[i].shardSetID).
+			SetShardFn(shardFn).
+			SetClientConnectionOptions(connectionOpts)
+		server := newTestServerSetup(t, serverOpts)
+		servers = append(servers, server)
+	}
+
+	// Start the servers.
+	log := xtest.NewLogger(t)
+	for i, server := range servers {
+		require.NoError(t, server.startServer())
+		log.Sugar().Infof("server %d is now up", i)
+	}
+
+	// Create clients for writing to the servers.
+	clients := make([]*client, 0, len(servers))
+	for _, server := range servers {
+		client := server.newClient()
+		require.NoError(t, client.connect())
+		clients = append(clients, client)
+	}
+
+	for _, server := range servers {
+		require.NoError(t, server.waitUntilLeader())
+	}
+
+	var (
+		idPrefix = "metric.id"
+		numIDs   = 100
+
+		start1    = getNowFn()
+		stop1     = start1.Add(10 * time.Second)
+		start2    = stop1.Add(time.Minute + 2*time.Second)
+		stop2     = start2.Add(10 * time.Second)
+		finalTime = stop2.Add(time.Minute + 2*time.Second)
+		interval  = 2 * time.Second
+
+		sleepDuration = time.Second
+	)
+	ids := generateTestIDs(idPrefix, numIDs)
+	testTimedMetadataTemplate := metadata.TimedMetadata{
+		AggregationID: maggregation.MustCompressTypes(maggregation.Sum),
+		StoragePolicy: policy.NewStoragePolicy(2*time.Second, xtime.Second, time.Hour),
+	}
+	metadataFn := func(idx int) metadataUnion {
+		timedMetadata := testTimedMetadataTemplate
+		return metadataUnion{
+			mType:         timedMetadataType,
+			timedMetadata: timedMetadata,
+		}
+	}
+	datasets := []testDataset{
+		mustGenerateTestDataset(t, datasetGenOpts{
+			start:        start1,
+			stop:         stop1,
+			interval:     interval,
+			ids:          ids,
+			category:     timedMetric,
+			typeFn:       roundRobinMetricTypeFn,
+			valueGenOpts: defaultValueGenOpts,
+			metadataFn:   metadataFn,
+		}),
+		mustGenerateTestDataset(t, datasetGenOpts{
+			start:        start2,
+			stop:         stop2,
+			interval:     interval,
+			ids:          ids,
+			category:     timedMetric,
+			typeFn:       roundRobinMetricTypeFn,
+			valueGenOpts: defaultValueGenOpts,
+			metadataFn:   metadataFn,
+		}),
+	}
+
+	for _, data := range datasets[0] {
+		setNowFn(data.timestamp)
+
+		for _, mm := range data.metricWithMetadatas {
+			idx := getServerIndex(mm.metric.ID(), initialPlacement)
+			require.NoError(t, clients[idx].writeTimedMetricWithMetadata(mm.metric.timed, mm.metadata.timedMetadata))
+		}
+		for _, c := range clients {
+			require.NoError(t, c.flush())
+		}
+
+		// Give server some time to process the incoming packets.
+		time.Sleep(sleepDuration)
+	}
+
+	setNowFn(start2)
+	time.Sleep(sleepDuration)
+	require.NoError(t, setPlacement(placementKey, kvStore, finalPlacement))
+	time.Sleep(sleepDuration)
```

Comment:
There is no guarantee that the placement update has actually happened after time x.
Maybe there is a way to query the nodes and see if they have the new placement applied, and then wait (with a timeout) on this condition?

Reply:
I didn't find a way to query placement status of aggregator. I could query `placementManager` for it, though it wouldn't prove that aggregator processed it yet.

At least with current implementation, placement updates should be processed in no time, since `placementTick()` gets notified about placement changes directly from `placementManager` and since tests use in-memory KV store implementation, which immediately calls watchers after each update.

https://github.com/m3db/m3/blob/f75f3e204da42b2c79ff2580c2fc2121cc2fcb85/src/aggregator/aggregator/aggregator.go#L184-L195

I guess I could add a way to view aggregators placements, or a method to force aggregator to process the most recent placement, though I'm not sure if it's worth it in this case.

Reply:
Well, there is no such thing as "immediate" in asynchronous code.
But I guess we can leave this as is for now, and get back to it after we implement the new "can lead" endpoint, which might turn out to be useful here.